pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,576 @@
+package purger
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/purger/purgeplan"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const millisecondPerDay = int64(24 * time.Hour / time.Millisecond)
+
+var separatorString = string([]byte{model.SeparatorByte})
+
+type deleteRequestWithLogger struct {
+	chunk.DeleteRequest
+	logger log.Logger
+}
+
+// DataPurgerConfig holds config for DataPurger
+type DataPurgerConfig struct {
+	EnablePurger    bool   `yaml:"enable_purger"`
+	NumWorkers      int    `yaml:"num_workers"`
+	ObjectStoreType string `yaml:"object_store_type"`
+}
+
+// RegisterFlags registers CLI flags for DataPurgerConfig
+func (cfg *DataPurgerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.EnablePurger, "purger.enable-purger", false, "Enable purger to allow deletion of series. Be aware that Delete series feature is still experimental")
+	f.IntVar(&cfg.NumWorkers, "purger.num-workers", 2, "Number of workers executing delete plans in parallel")
+	f.StringVar(&cfg.ObjectStoreType, "purger.object-store-type", "", "Name of the object store to use for storing delete plans")
+}
+
+type workerJob struct {
+	planNo          int
+	userID          string
+	deleteRequestID string
+	logger          log.Logger
+}
+
+type workerJobExecutionStatus struct {
+	workerJob
+	err error
+}
+
+// DataPurger does the purging of data which is requested to be deleted
+type DataPurger struct {
+	cfg           DataPurgerConfig
+	deleteStore   *chunk.DeleteStore
+	chunkStore    chunk.Store
+	storageClient chunk.ObjectClient
+
+	executePlansChan             chan deleteRequestWithLogger
+	workerJobChan                chan workerJob
+	workerJobExecutionStatusChan chan workerJobExecutionStatus
```

Comment:
I don't see why `workerJobExecutionStatusChan` is needed at all. Same cleanup can be done by worker itself, after it is finished with the plan? Why does it need to be done on separate goroutine?

Furthermore, as it is now, when Stop is called, goroutine reading from `workerJobExecutionStatusChan` will exit quickly (because it's not doing anything that takes a long time), while workers are still executing the plan, but when they are finished, there is no channel reader anymore, so the result of plan execution will not be logged or handled in any way.

Reply:
While there can be multiple plans, `worker` just takes care of just 1 plan so there has to be another entity which does the cleanup after all the plans are executed successfully. I will take care of the exit issue that you pointed out.

Reply:
> While there can be multiple plans, `worker` just takes care of just 1 plan so there has to be another entity which does the cleanup after all the plans are executed successfully.

I understand this, but think that worker can simply check if it did the last piece on its own, without going through another goroutine.

Reply:
Since we are building per day plan, the last one could be smallest and could finish quickly which means we might mark a delete request completed before it executes all the plans successfully.
We also do not want to mark a delete request completed when any of its plans but the last one fails to execute.

Reply:
That's fine... I think your checks are fine, I'm just objecting on using another goroutine and extra channel. Worker can do the same checks, and remove entry from the map only if there no other plan that runs. If you extract that code to new method as suggested elsewhere, I would just let worker call the method, instead of spawning new goroutine.

Reply:
Sorry, I got your point. I have done the changes and pushed them. Thanks!